<launch>

    <!-- Choose visualization -->
    <arg name="rviz" default="true"/>
    <arg name="rtabmapviz" default="false"/>

    <param name="use_sim_time" type="bool" value="true"/>

    <!-- OpenCV StereoSGBM Mapping Node -->
    <node name="stereoSGBM_mapping" pkg="stereoSGBM_mapping" type="stereoSGBM_mapping" respawn="false" output="screen">
        <param name="capture_width"     type="int"      value="640"/>
        <param name="capture_height"    type="int"      value="480"/>
        <param name="capture_fps"       type="int"      value="20"/>
        <param name="calibration_file"  type="string"   value="$(find stereoSGBM_mapping)calibration/stereo_calibration.xml"/>
    </node>

    <!-- ICM20948 IMU Odometry -->
    <node name="icm20948_odometry" pkg="icm20948_odometry" type="icm20948_odometry" respawn="false" output="screen">
    </node>

    <!-- Localization-only mode -->
    <arg name="localization" default="false"/>
    <arg if="$(arg localization)" name="rtabmap_args" default=""/>
    <arg unless="$(arg localization)" name="rtabmap_args" default="--delete_db_on_start"/>

    <group ns="rtabmap">
        <!-- SLAM -->
        <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
            <param name="frame_id"              type="string"   value="base_footprint"/>
            <param name="wait_for_transform"    type="bool"     value="true"/>

            <param name="subscribe_depth"       type="bool"     value="true"/>
            <param name="subscribe_scan"        type="bool"     value="false"/>

            <!-- odometry? -->

            <remap from="rgb/image"         to="/data_throttled_image"/>
            <remap from="depth/image"       to="/data_throttled_image_depth"/>
            <remap from="rgb/camera_info"   to="/data_throttled_camera_info"/>

            <param name="rgb/image_transport"       type="string" value="compressed"/>
            <param name="depth/image_transport"     type="string" value="compressedDepth"/>

            <!-- RTAB-Map's parameters: do "rosrun rtabmap rtabmap (double-dash)params" to see list of available parameters -->
            <param name="RGBD/NeighborLinkRefining"         type="string" value="true"/>
            <param name="RGBD/ProximityBySpace"             type="string" value="true"/>
            <param name="RGBD/ProximityByTime"              type="string" value="false"/>
            <param name="RGBD/ProximityPathMaxNeighbors"    type="string" value="10"/>
            <param name="RGBD/OptimizeFromGraphEnd"         type="string" value="false"/>
            <param name="RGBD/OptimizeMaxError"             type="string" value="4"/>
            <param name="RGBD/LocalRadius"                  type="string" value="5"/>
            <param name="Reg/Strategy"                      type="string" value="1"/>
            <param name="Reg/Force3DoF"                     type="string" value="true"/>
            <param name="Vis/MinInliers"                    type="string" value="12"/>
            <param name="Grid/FromDepth"                    type="string" value="false"/>
            <param name="Mem/STMSize"                       type="string" value="30"/>
            <param name="Icp/CorrespondenceRatio"           type="string" value="0.2"/>
            <param name="Icp/PM"                            type="string" value="false"/>
            <param name="Icp/PointToPlane"                  type="string" value="false"/>
            <param name="Icp/MaxCorrespondenceDistance"     type="string" value="0.15"/>
            <param name="Icp/VoxelSize"                     type="string" value="0.05"/>

            <!-- Localization mode -->
            <param if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
            <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
            <param name="Mem/InitWMWithAllnodes" tpye="string" value="$(arg localization)"/>
        </node>

        <!-- Visualisation RTAB-Map -->
        <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
            <param name="subscribe_depth"       type="bool"     value="true"/>
            <param name="subscribe_scan"        type="bool"     value="false"/>
            <param name="frame_id"              type="string"   value="base_footprint"/>
            <param name="wait_for_transform"    type="bool"     value="true"/>

            <remap from="rgb/image"         to="/data_throttled_image"/>
            <remap from="depth/image"       to="/data_throttled_image_depth"/>
            <remap from="rgb/camera_info"   to="/data_throttled_camera_info"/>
            <remap from="odom"              to="/az3/base_controller/odom"/>

            <param name="rgb/image_transport"   type="string"   value="compressed"/>
            <param name="depth/image_transport" type="string"   value="compressedDepth"/>
        </node>
    </group>

    <!-- RViz Visualization -->
    <node if="$(arg rviz)" name="rviz" pkg="rviz" type="rviz" args="-d $(find rtabmap_ros)/launch/config/demo_robot_mapping.rviz" output="screen"/>
        <remap from="rgb/image"         to="/data_throttled_image"/>
        <remap from="depth/image"       to="/data_throttled_image_depth"/>
        <remap from="rgb/camera_info"   to="/data_throttled_camera_info"/>

        <param name="rgb/image_transport"   type="string"   value="compressed"/>
        <param name="depth/image_transport" type="string"   value="compressedDepth"/>
    </node>

</launch>